File: src/main/java/juuxel/lakeside/biome/MoreOverworldBiomes.java
package juuxel.lakeside.biome;

import com.google.common.collect.Multimap;
import com.google.common.collect.MultimapBuilder;
import net.minecraft.util.registry.BuiltinRegistries;
import net.minecraft.util.registry.RegistryKey;
import net.minecraft.world.biome.Biome;
import net.minecraft.world.biome.layer.util.LayerRandomnessSource;

public final class MoreOverworldBiomes {
    private static final Multimap<Integer, SmallVariantEntry> smallVariants = MultimapBuilder.hashKeys().hashSetValues().build();
    private static final Multimap<Integer, SmallVariantEntry> islands = MultimapBuilder.hashKeys().hashSetValues().build();

    public static void addSmallVariant(RegistryKey<Biome> base, RegistryKey<Biome> variant, int chance) {
        smallVariants.put(getRawId(base), new SmallVariantEntry(getRawId(variant), chance));
    }

    public static int transformSmallVariant(int base, LayerRandomnessSource random) {
        return transform(smallVariants, base, random);
    }

    public static void addIsland(RegistryKey<Biome> base, RegistryKey<Biome> variant, int chance) {
        islands.put(getRawId(base), new SmallVariantEntry(getRawId(variant), chance));
    }

    public static int transformIsland(int base, LayerRandomnessSource random) {
        return transform(islands, base, random);
    }

    private static int transform(Multimap<Integer, SmallVariantEntry> variants, int base, LayerRandomnessSource random) {
        if (variants.containsKey(base)) {
            var entries = variants.get(base);
            var bound = 0;

            for (var entry : entries) {
                bound += entry.chance();
            }

            var i = random.nextInt(bound);
            var offset = 0;

            for (var entry : entries) {
                if (i == offset) return entry.variant();
                offset += entry.chance();
            }
        }

        return base;
    }

    private static int getRawId(RegistryKey<Biome> key) {
        return BuiltinRegistries.BIOME.getRawId(BuiltinRegistries.BIOME.getOrThrow(key));
    }

    private record SmallVariantEntry(int variant, int chance) {}
}

File: src/main/kotlin/juuxel/lakeside/biome/MoreOverworldBiomes.kt
package juuxel.lakeside.biome

import com.google.common.collect.Multimap
import com.google.common.collect.MultimapBuilder
import net.minecraft.util.registry.BuiltinRegistries
import net.minecraft.util.registry.RegistryKey
import net.minecraft.world.biome.Biome
import net.minecraft.world.biome.layer.util.LayerRandomnessSource

@Suppress("UnstableApiUsage")
object MoreOverworldBiomes {
    private val smallVariants: Multimap<Int, SmallVariantEntry> = MultimapBuilder.hashKeys().hashSetValues().build()
    private val islands: Multimap<Int, SmallVariantEntry> = MultimapBuilder.hashKeys().hashSetValues().build()

    fun addSmallVariant(base: RegistryKey<Biome>, variant: RegistryKey<Biome>, chance: Int) {
        smallVariants.put(getRawId(base), SmallVariantEntry(getRawId(variant), chance))
    }

    fun transformSmallVariant(base: Int, random: LayerRandomnessSource): Int =
        transform(smallVariants, base, random)

    fun addIsland(base: RegistryKey<Biome>, variant: RegistryKey<Biome>, chance: Int) {
        islands.put(getRawId(base), SmallVariantEntry(getRawId(variant), chance))
    }

    fun transformIsland(base: Int, random: LayerRandomnessSource): Int = transform(islands, base, random)

    private fun transform(
        variants: Multimap<Int, SmallVariantEntry>, base: Int, random: LayerRandomnessSource
    ): Int {
        if (variants.containsKey(base)) {
            val entries = variants[base]
            val bound = entries.sumBy { it.chance }
            val i = random.nextInt(bound)
            var offset = 0

            for (entry in entries) {
                if (i == offset) return entry.variant
                offset += entry.chance
            }
        }

        return base
    }

    private fun getRawId(key: RegistryKey<Biome>): Int =
        BuiltinRegistries.BIOME.getRawId(BuiltinRegistries.BIOME.getOrThrow(key))

    private data class SmallVariantEntry(val variant: Int, val chance: Int)
}
